dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -2125,5 +2149,108 @@ bool Segment::placeDelete(const DMContext & dm_context,
+    auto bitmap_filter = std::make_shared<BitmapFilter>(total_rows, segment_snap->clone());
+    for (;;)
+    {
+        auto blk = stream->read();
+        if (likely(blk))
+        {
+            bitmap_filter->set(blk.segmentRowIdCol());
+        }
+        else
+        {
+            break;
+        }
+    }
+    bitmap_filter->runOptimize();
+    LOG_DEBUG(log, "buildBitmapFilter total_rows={} cost={}ms", total_rows, sw_total.elapsedMilliseconds());
+    return bitmap_filter;
+}
+
+BlockInputStreamPtr Segment::getBitmapFilterInputStream(BitmapFilterPtr && bitmap_filter,
+                                                        const DMContext & dm_context,
+                                                        const ColumnDefines & columns_to_read,
+                                                        const RowKeyRanges & data_ranges,
+                                                        const RSOperatorPtr & filter,
+                                                        UInt64 max_version,
+                                                        size_t expected_block_size)
+{
+    auto segment_snap = bitmap_filter->snapshot();
+    BlockInputStreamPtr stable_stream = segment_snap->stable->getInputStream(
+        dm_context,
+        columns_to_read,
+        data_ranges,
+        filter,
+        max_version,
+        expected_block_size,
+        /* enable_clean_read */ false,
+        /* is_fast_scan */ false);
+
+    auto columns_to_read_ptr = std::make_shared<ColumnDefines>(columns_to_read);
+    // TODO: Scanning the filter and determine which block to read instead of read block and filter.
+    BlockInputStreamPtr delta_stream = std::make_shared<DeltaValueInputStream>(
+        dm_context,
+        segment_snap->delta,
+        columns_to_read_ptr,
+        this->rowkey_range);
+
+    return std::make_shared<BitmapFilterBlockInputStream>(
+        columns_to_read,
+        stable_stream,
+        delta_stream,
+        segment_snap->stable->getDMFilesRows(),
+        segment_snap->delta->getRows(),
+        bitmap_filter,
+        dm_context.tracing_id);
+}
+
+BlockInputStreamPtr Segment::getBitmapFilterInputStream(const DMContext & dm_context,
+                                                        const ColumnDefines & columns_to_read,
+                                                        const SegmentSnapshotPtr & segment_snap,
+                                                        const RowKeyRanges & data_ranges,
+                                                        const RSOperatorPtr & filter,
+                                                        UInt64 max_version,
+                                                        size_t expected_block_size)
+{
+    auto bitmap_filter = buildBitmapFilter(
+        dm_context,
+        segment_snap,
+        data_ranges,
+        filter,
+        max_version,
+        expected_block_size);
```

Comment:
Do we need to build bitmap filter each query? Maybe just one bitmap for one segment snapshot?

Reply:
accidentally approved……